src/twisted/internet/defer.py

```diff
@@ -217,10 +217,10 @@ def maybeDeferred(
     except BaseException:
         return fail(Failure(captureVars=Deferred.debug))
 
-    if type(result) in _DEFERRED_SUBCLASSES:
+    if hasattr(result, "_twistedPrivateIsDeferred"):
```

Comment:
I worry about this particular optimization. `Deferred`s can have _any_ object, and potentially `__getattr__` may do unexpected things on arbitrary objects, so looking up attributes on them will, in rare cases, break things.

Reply:
An alternative might be to have a Cython extension for is-it-a-Deferred; you'd switch from a single C call (`type()`) to a single C call, but then the implementation even if the same approach (global vec of known types) could be faster. Rather more implementation overhead, though.

Reply:
I'm also kind of concerned about this one, but I think it might still be worthwhile; however, could we push this one off to a separate PR to avoid all these other optimizations getting bogged down in controversy and fretting about it?

Reply:
Oh wait, this is the whole thing, isn't it :) hmm…

Reply:
How bad is it to call `isinstance` if your type has an optimized `__instancecheck__`, I wonder…?

Reply:
I believe this would be even slower than regular `isinstance`, because it is implemented in C and we would introducing another place to call python code.

`hasattr` must be the most simple operation.